src/python/pants/rules/core/fmt.py

```diff
@@ -29,8 +29,15 @@ def from_execute_process_result(process_result: ExecuteProcessResult) -> "FmtRes
     )
 
 
-class FmtResults(Collection[FmtResult]):
-  """This collection allows us to aggregate multiple `FmtResult`s for a language."""
+@dataclass(frozen=True)
+class AggregatedFmtResults:
+  """This collection allows us to safely aggregate multiple `FmtResult`s for a language.
+
+  The `combined_digest` is used to ensure that none of the formatters overwrite each other. The
+  language implementation should run each formatter one at a time and pipe the resulting digest of
+  one formatter into the next."""
+  results: Tuple[FmtResult, ...]
+  combined_digest: Optional[Digest]
```

Comment:
Why is this optional? What does it being empty signify?

Reply:
Good question. I was thinking it handles the case where you have zero formatters installed for that language. Instead, we should maybe have the language set the value to either EMPTY_DIGEST or the original target’s digest.

Reply:
Thanks for flagging this. I realized that it's wrong to make this Optional because it means that the directory merge would fail when the `combined_digest=None`. Instead, a no-op should be the original digest for the target, as you state in your suggested change for the docstring. This now makes sense to me. Thanks!